File: minting/src/main/scala/co/topl/minting/interpreters/OperationalKeyMaker.scala
package co.topl.minting.interpreters

import cats.Traverse
import cats.data._
import cats.effect.Deferred
import cats.effect.implicits._
import cats.effect.{Async, MonadCancelThrow, Ref, Resource, Sync}
import cats.implicits._
import co.topl.algebras.ClockAlgebra.implicits._
import co.topl.algebras._
import co.topl.codecs.bytes.tetra.instances._
import co.topl.consensus.algebras.LeaderElectionValidationAlgebra
import co.topl.consensus.algebras.ConsensusValidationStateAlgebra
import co.topl.crypto.generation.mnemonic.Entropy
import co.topl.crypto.signing._
import co.topl.minting.algebras._
import co.topl.minting.models.OperationalKeyOut
import co.topl.models._
import co.topl.models.utility._
import co.topl.consensus.models._
import co.topl.crypto.models.SecretKeyKesProduct
import co.topl.typeclasses.implicits._
import com.google.common.primitives.Longs
import com.google.protobuf.ByteString
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.Slf4jLogger

import java.util.UUID

object OperationalKeyMaker {

  /**
   * Constructs an OperationalKeys interpreter using a SecureStore.
   *
   * @param activationOperationalPeriod The operational period number in which the staker becomes active
   * @param address                     The staker's address
   * @param etaCalculation An EtaCalculation interpreter is needed to determine the eta to use when determining VRF ineligibilities.
   * @param consensusState Used for the lookup of relative stake for VRF ineligibilities
   */
  def make[F[_]: Async](
    activationOperationalPeriod: Long,
    address:                     StakingAddress,
    vrfConfig:                   VrfConfig,
    secureStore:                 SecureStore[F],
    clock:                       ClockAlgebra[F],
    vrfCalculator:               VrfCalculatorAlgebra[F],
    leaderElection:              LeaderElectionValidationAlgebra[F],
    consensusState:              ConsensusValidationStateAlgebra[F],
    kesProductResource:          Resource[F, KesProduct],
    ed25519Resource:             Resource[F, Ed25519]
  ): Resource[F, OperationalKeyMakerAlgebra[F]] =
    for {
      // Delay further initialization until the activation period starts
      currentOperationalPeriod <- clock.globalOperationalPeriod.toResource
      _ <- Async[F]
        .whenA(activationOperationalPeriod > currentOperationalPeriod)(
          clock.operationalPeriodRange(activationOperationalPeriod).map(_.start).flatMap(clock.delayedUntilSlot)
        )
        .toResource
      stateRef                     <- Ref.of(none[(Long, Map[Long, Deferred[F, Option[OperationalKeyOut]]])]).toResource
      implicit0(logger: Logger[F]) <- Slf4jLogger.fromName("OperationalKeyMaker").toResource
      impl = new Impl[F](
        activationOperationalPeriod,
        address,
        vrfConfig,
        secureStore,
        clock,
        vrfCalculator,
        leaderElection,
        consensusState,
        kesProductResource,
        ed25519Resource,
        stateRef
      )
    } yield impl

  private class Impl[F[_]: Async: Logger](
    activationOperationalPeriod: Long,
    address:                     StakingAddress,
    vrfConfig:                   VrfConfig,
    secureStore:                 SecureStore[F],
    clock:                       ClockAlgebra[F],
    vrfCalculator:               VrfCalculatorAlgebra[F],
    leaderElection:              LeaderElectionValidationAlgebra[F],
    consensusState:              ConsensusValidationStateAlgebra[F],
    kesProductResource:          Resource[F, KesProduct],
    ed25519Resource:             Resource[F, Ed25519],
    stateRef:                    Ref[F, Option[(Long, Map[Long, Deferred[F, Option[OperationalKeyOut]]])]]
  ) extends OperationalKeyMakerAlgebra[F] {

    def operationalKeyForSlot(slot: Slot, parentSlotId: SlotId, eta: Eta): F[Option[OperationalKeyOut]] =
      clock
        .operationalPeriodOf(slot)
        .flatMap(operationalPeriod =>
          MonadCancelThrow[F].uncancelable(_ =>
            stateRef.get.flatMap {
              case Some((`operationalPeriod`, keys)) =>
                OptionT.fromOption[F](keys.get(slot)).flatMapF(_.get).value
              case _ =>
                OptionT(consensusState.operatorRelativeStake(parentSlotId.blockId, slot)(address))
                  .flatMapF(relativeStake =>
                    consumeEvolvePersist(
                      (operationalPeriod - activationOperationalPeriod).toInt,
                      slot,
                      relativeStake,
                      eta
                    )
                  )
                  .semiflatTap(newKeys => stateRef.set((operationalPeriod -> newKeys).some))
                  .flatTapNone(stateRef.set(none))
                  .subflatMap(_.get(slot))
                  .flatMapF(_.get)
                  .value
            }
          )
        )

    /**
     * Consume the current key from disk.  Exactly one key is expected; if 0 or more than is detected, an error is raised.
     * If they key is behind the `currentOperationalPeriod`, it is first updated to the `currentOperationalPeriod`.  The
     * key for `timeStep + 1` is constructed and saved to disk.
     */
    private[interpreters] def consumeEvolvePersist(
      timeStep:      Int,
      slot:          Slot,
      relativeStake: Ratio,
      eta:           Eta
    ): F[Option[Map[Slot, Deferred[F, Option[OperationalKeyOut]]]]] =
      MonadCancelThrow[F].uncancelable(_ =>
        (
          for {
            fileName <- OptionT(
              secureStore.list
                .ensure(new IllegalStateException("SecureStore is empty"))(_.nonEmpty)
                .ensure(new IllegalStateException("SecureStore contains multiple keys"))(_.length === 1)
                .map(_.headOption)
            )
            _       <- OptionT.liftF(Logger[F].info(show"Consuming key id=$fileName"))
            diskKey <- OptionT(secureStore.consume[SecretKeyKesProduct](fileName))
            latest <- OptionT.liftF(
              kesProductResource.use(kesProduct => Sync[F].delay(kesProduct.getCurrentStep(diskKey)))
            )
            currentPeriodKey <-
              if (latest === timeStep) OptionT.pure[F](diskKey)
              else if (latest > timeStep)
                OptionT
                  .none[F, SecretKeyKesProduct]
                  .flatTapNone(
                    Logger[F].info(
                      show"Persisted key timeStep=$latest is greater than current timeStep=$timeStep." +
                      show"  Re-persisting original key."
                    ) >>
                    secureStore.write(fileName, diskKey)
                  )
              else
                OptionT.liftF(kesProductResource.use(kesProduct => Sync[F].delay(kesProduct.update(diskKey, timeStep))))
            nextTimeStep = timeStep + 1
            onComplete = for {
              _ <- Sync[F].defer(Logger[F].info(show"Saving next key idx=$nextTimeStep"))
              updated <-
                kesProductResource.use(kesProduct => Sync[F].delay(kesProduct.update(currentPeriodKey, nextTimeStep)))
              _ <- secureStore.write(UUID.randomUUID().toString, updated)
            } yield ()
            res <- OptionT.liftF(
              prepareOperationalPeriodKeys(currentPeriodKey, slot, relativeStake, eta)(onComplete)
            )
          } yield res
        ).value
      )

    /**
     * Using some KES parent, construct the linear keys for the upcoming operational period.  A linear key is constructed
     * for each slot for which we _might_ be eligible for VRF.
     */
    private[interpreters] def prepareOperationalPeriodKeys(
      parentSK:      SecretKeyKesProduct,
      fromSlot:      Slot,
      relativeStake: Ratio,
      eta:           Eta
    )(onComplete: => F[Unit]): F[Map[Slot, Deferred[F, Option[OperationalKeyOut]]]] =
      for {
        epoch                  <- clock.epochOf(fromSlot)
        operationalPeriod      <- clock.operationalPeriodOf(fromSlot)
        operationalPeriodSlots <- clock.operationalPeriodRange(operationalPeriod).map(_.toList)
        _ <- Logger[F].info(
          show"Computing operational keys for" +
          show" epoch=$epoch" +
          show" eta=$eta" +
          show" range=${operationalPeriodSlots.head}..${operationalPeriodSlots.last}"
        )
        parentVK      <- kesProductResource.use(kesProduct => Sync[F].delay(kesProduct.getVerificationKey(parentSK)))
        deferredSlots <- operationalPeriodSlots.traverse(slot => Deferred[F, Option[OperationalKeyOut]].map((slot, _)))
        threshold     <- maximumThreshold(relativeStake)
        // Launch a background fiber which will create the child keys for the new operational period.
        // As each child key is created, the corresponding Deferred instance is completed.
        _ <- Async[F]
          .uncancelable(poll =>
            Async[F].cede >>
            // Allow cancelation of the child key creation
            poll(
              fulfillDeferredSlots(epoch, eta, threshold, parentSK, parentVK, operationalPeriodSlots)(
                deferredSlots
              ).void
            ) >>
            // But disable cancelation for the onComplete (evolve) action
            onComplete
          )
          .start
      } yield deferredSlots.toMap

    private def fulfillDeferredSlots[G[_]: Traverse](
      epoch:                  Epoch,
      eta:                    Eta,
      threshold:              Ratio,
      parentSK:               SecretKeyKesProduct,
      parentVK:               VerificationKeyKesProduct,
      operationalPeriodSlots: Iterable[Slot]
    )(deferredSlots: G[(Slot, Deferred[F, Option[OperationalKeyOut]])]) =
      deferredSlots.traverse { case (slot, deferred) =>
        potentiallyEligibleSlot(eta, slot, threshold)
          .ifM(
            ifTrue = prepareOperationalPeriodKey(parentSK, parentVK, slot).map(_.some),
            ifFalse = none[OperationalKeyOut].pure[F]
          )
          .flatMap(deferred.complete)
      } >> Logger[F].info(
        show"Finished computing operational keys for" +
        show" epoch=$epoch" +
        show" eta=$eta" +
        show" range=${operationalPeriodSlots.head}..${operationalPeriodSlots.last}"
      )

    /**
     * From some "parent" KES keypair, create a single child key for the given slot
     */
    private def prepareOperationalPeriodKey(
      parentSK: SecretKeyKesProduct,
      parentVK: VerificationKeyKesProduct,
      slot:     Slot
    ) =
      for {
        entropy      <- Sync[F].delay(Entropy.fromUuid(UUID.randomUUID()))
        childKeyPair <- ed25519Resource.use(ed => Sync[F].delay(ed.deriveKeyPairFromEntropy(entropy, None)))
        _            <- Async[F].cede
        message = childKeyPair.verificationKey.bytes ++ Longs.toByteArray(slot)
        parentSignature <- kesProductResource
          .use(kesProductScheme => Sync[F].delay(kesProductScheme.sign(parentSK, message)))
      } yield OperationalKeyOut(
        slot,
        ByteString.copyFrom(childKeyPair.verificationKey.bytes),
        ByteString.copyFrom(childKeyPair.signingKey.bytes),
        parentSignature,
        parentVK
      )

    /**
     * Determines the threshold at the VRF LDD Cutoff point.
     * @param relativeStake The operator's relative stake
     * @return a threshold at the LDD Cutoff Point
     */
    private def maximumThreshold(relativeStake: Ratio): F[Ratio] =
      leaderElection.getThreshold(relativeStake, vrfConfig.lddCutoff)

    /**
     * Determines if the operator *might* be eligible for the given slot,
     * based on the threshold at the LDD Cutoff.  If the operator is not
     * eligible at the cutoff slot, then it will never be eligible
     * @param eta Eta value
     * @param slot Test slot
     * @param threshold LDD Cutoff Threshold
     * @return true if potentially eligible, false if definitely ineligible
     */
    private def potentiallyEligibleSlot(eta: Eta, slot: Slot, threshold: Ratio) =
      vrfCalculator
        .rhoForSlot(slot, eta)
        .flatMap(leaderElection.isSlotLeaderForThreshold(threshold))
  }

}

File: dart/minting/lib/interpreters/operational_key_maker.dart
import 'dart:math';

import 'package:bifrost_codecs/codecs.dart';
import 'package:bifrost_common/algebras/clock_algebra.dart';
import 'package:bifrost_common/models/common.dart';
import 'package:bifrost_consensus/algebras/consensus_validation_state_algebra.dart';
import 'package:bifrost_consensus/algebras/eta_calculation_algebra.dart';
import 'package:bifrost_crypto/ed25519.dart' show ed25519;
import 'package:bifrost_crypto/kes.dart';
import 'package:bifrost_minting/algebras/operational_key_maker_algebra.dart';
import 'package:bifrost_minting/algebras/secure_store_algebra.dart';
import 'package:bifrost_minting/algebras/vrf_calculator_algebra.dart';
import 'package:bifrost_minting/interpreters/vrf_calculator.dart';
import 'package:fixnum/fixnum.dart';
import 'package:fpdart/fpdart.dart';
import 'package:logging/logging.dart';
import 'package:rational/rational.dart';
import 'package:topl_protobuf/consensus/models/operational_certificate.pb.dart';
import 'package:topl_protobuf/consensus/models/slot_data.pb.dart';
import 'package:topl_protobuf/consensus/models/staking_address.pb.dart';
import 'package:topl_protobuf/quivr/models/shared.pb.dart';

class OperationalKeyMaker extends OperationalKeyMakerAlgebra {
  final Int64 operationalPeriodLength;
  final Int64 activationOperationalPeriod;
  final StakingAddress address;
  final SecureStoreAlgebra secureStore;
  final ClockAlgebra clock;
  final VrfCalculatorAlgebra vrfCalculator;
  final EtaCalculationAlgebra etaCalculation;
  final ConsensusValidationStateAlgebra consensusValidationState;
  Int64 currentOperationalPeriod;
  Map<Int64, OperationalKeyOut>? currentKeyCache;

  final log = Logger("OperationalKeyMaker");

  OperationalKeyMaker(
    this.operationalPeriodLength,
    this.activationOperationalPeriod,
    this.address,
    this.secureStore,
    this.clock,
    this.vrfCalculator,
    this.etaCalculation,
    this.consensusValidationState,
    this.currentOperationalPeriod,
    this.currentKeyCache,
  );

  static Future<OperationalKeyMaker> init(
    SlotId parentSlotId,
    Int64 operationalPeriodLength,
    Int64 activationOperationalPeriod,
    StakingAddress address,
    SecureStoreAlgebra secureStore,
    ClockAlgebra clock,
    VrfCalculatorAlgebra vrfCalculator,
    EtaCalculationAlgebra etaCalculation,
    ConsensusValidationStateAlgebra consensusValidationState,
    SecretKeyKesProduct initialSK,
  ) async {
    Int64 slot = clock.globalSlot;
    if (slot < 0) slot = Int64.ZERO;
    final operationalPeriod = slot ~/ operationalPeriodLength;
    final impl = OperationalKeyMaker(
      operationalPeriodLength,
      activationOperationalPeriod,
      address,
      secureStore,
      clock,
      vrfCalculator,
      etaCalculation,
      consensusValidationState,
      operationalPeriod,
      null,
    );

    await secureStore.write("k", initialSK.encode);

    final relativeStake = await consensusValidationState.operatorRelativeStake(
        parentSlotId.blockId, slot, address);
    if (relativeStake != null) {
      final initialKeysOpt = await impl._consumeEvolvePersist(
          (operationalPeriod - activationOperationalPeriod).toInt(),
          (t) => impl._prepareOperationalPeriodKeys(
              t, slot, parentSlotId, relativeStake));
      impl.currentKeyCache = initialKeysOpt;
      impl.currentOperationalPeriod = operationalPeriod;
    }
    return impl;
  }

  @override
  Future<OperationalKeyOut?> operationalKeyForSlot(
      Int64 slot, SlotId parentSlotId) async {
    final operationalPeriod = slot ~/ operationalPeriodLength;
    if (operationalPeriod == currentOperationalPeriod)
      return currentKeyCache?[slot];
    final relativeStake = await consensusValidationState.operatorRelativeStake(
        parentSlotId.blockId, slot, address);
    if (relativeStake == null) return null;
    final newKeys = await _consumeEvolvePersist(
        (operationalPeriod - activationOperationalPeriod).toInt(),
        (t) => _prepareOperationalPeriodKeys(
            t, slot, parentSlotId, relativeStake));
    currentOperationalPeriod = operationalPeriod;
    currentKeyCache = newKeys;
    return newKeys?[slot];
  }

  Future<T?> _consumeEvolvePersist<T>(
      int timeStep, Future<T> Function(SecretKeyKesProduct) use) async {
    final fileNames = await secureStore.list();
    if (fileNames.length != 1)
      throw Exception("SecureStore contained invalid number of keys");
    final fileName = fileNames.first;
    log.info("Consuming key id=$fileName");
    final diskKeyBytes = await secureStore.consume(fileName);
    if (diskKeyBytes == null) return null;
    final SecretKeyKesProduct diskKey =
        SecretKeyKesProduct.decode(diskKeyBytes);
    final latest = await kesProduct.getCurrentStep(diskKey);
    SecretKeyKesProduct? currentPeriodKey;
    if (latest == timeStep)
      currentPeriodKey = diskKey;
    else if (latest > timeStep) {
      log.info(
          "Persisted key timeStep=$latest is greater than current timeStep=$timeStep." +
              "  Re-persisting original key.");
      secureStore.write(fileName, diskKeyBytes);
    } else {
      currentPeriodKey = await kesProduct.update(diskKey, timeStep);
    }

    if (currentPeriodKey == null) return null;
    final res = await use(currentPeriodKey);
    final nextTimeStep = timeStep + 1;
    log.info("Updating next key idx=$nextTimeStep");
    final updated = await kesProduct.update(currentPeriodKey, nextTimeStep);
    log.info("Saving next key idx=$nextTimeStep");
    await secureStore.write("k", updated.encode);
    log.info("Saved next key idx=$nextTimeStep");
    return res;
  }

  Future<Map<Int64, OperationalKeyOut>> _prepareOperationalPeriodKeys(
      SecretKeyKesProduct kesParent,
      Slot fromSlot,
      SlotId parentSlotId,
      Rational relativeStake) async {
    final epoch = clock.epochOfSlot(fromSlot);
    final eta = await etaCalculation.etaToBe(parentSlotId, fromSlot);
    final operationalPeriod = fromSlot ~/ operationalPeriodLength;

    final operationalPeriodSlotMin =
        operationalPeriod * operationalPeriodLength;
    final operationalPeriodSlotMax =
        (operationalPeriod + 1) * operationalPeriodLength;
    log.info("Computing ineligible slots for" +
        " epoch=$epoch" +
        " eta=${eta.show}" +
        " range=$operationalPeriodSlotMin..$operationalPeriodSlotMax");
    final ineligibleSlots = [];
    // TODO: Uncomment once VRF is fixed
    // final ineligibleSlots = await vrfCalculator.ineligibleSlots(
    //     epoch,
    //     eta,
    //     Tuple2(operationalPeriodSlotMin, operationalPeriodSlotMax),
    //     relativeStake);
    final slots = List.generate(
            (operationalPeriodLength - (fromSlot % operationalPeriodLength))
                .toInt(),
            (i) => fromSlot + i)
        .where((s) => !ineligibleSlots.contains(s))
        .toList();
    log.info("Preparing linear keys. count=${slots.length}");
    final outs = await _prepareOperationalPeriodKeysImpl(kesParent, slots);
    final mappedKeys =
        Map.fromEntries(outs.map((out) => MapEntry(out.slot, out)));
    return mappedKeys;
  }

  Future<List<OperationalKeyOut>> _prepareOperationalPeriodKeysImpl(
      SecretKeyKesProduct kesParent, List<Slot> slots) async {
    final result = <OperationalKeyOut>[];
    final parentVK = await kesProduct.generateVerificationKey(kesParent);
    for (int i = 0; i < slots.length; i++) {
      final slot = slots[i];
      final childKeyPair = await ed25519.generateKeyPair();
      final parentSignature = await kesProduct.sign(
          kesParent,
          <int>[]
            ..addAll(childKeyPair.vk)
            ..addAll(slot.immutableBytes));
      result.add(
          OperationalKeyOut(slot, childKeyPair, parentSignature, parentVK));
    }
    return result;
  }
}
